### ISSUE TITLE
Allow many-to-one mappings

### ISSUE THREAD

**Issue by vincentarelbundock** [vincentarelbundock/countrycode#66]
This is a major refactoring project (unlikely to happen very soon). Not clear yet how to implement this. Post ideas below.

Use case: 

https://github.com/vincentarelbundock/countrycode/issues/61

https://github.com/vincentarelbundock/countrycode/issues/62

**Comment by LaDilettante** [comment]
I think the solution to this is to go with a relational database, with the primary key being the country ID (i.e. a sort of Platonic identity of that country).

Then we'll have table `countryname`, table `ISO`, etc. with the countryID as the foreign key in each of those table.

For example, let's say the ID for Kosovo is 100, then in table `ISO3` we have

| foreign_key | ISO3 |
| ---------------| ------- | 
| 100        | XKX |
| 100 | XKS |
| 100 | XKV |

Basically, a many-to-many SQL pattern.

Is this something you would like to see a PR? It would indeed be an almost complete re-write.

**Comment by vincentarelbundock** [comment]
I'm hesitant to say "yes" because I've found that even reviewing reasonably clean PRs requires a significant investment in time, and I have basically zero time to invest.

What your suggesting essentially sounds like a collection of many to one dictionaries. Io m worried that this may become difficult to update and maintain. It also doesn't solve the panel issue with different codes for different years. I also hesitate to motivate a full rewrite based on corner cases that are not actually supported codes by the organization (see Closed issue where we discuss Kosovo).

One intermediary step might be to maintain a few dataframes of corner cases. For instance, if you do a full join of your Kosovo example with countrycode::countrycode_data and feed that to the new custom dictionary argument (available in the GitHub version only), then country code will already give you many to one conversion.

**Comment by LaDilettante** [comment]
Thanks for the reply -- I'm sure you know best about the package and its use case, so I'll defer to your judgement. Thanks for a great package that I use everyday.

**Comment by vincentarelbundock** [comment]
Yeah, sorry, I didn't mean to shut you down. I just need to be convinced there's a considerable payoff, since I perceive the costs of re-factor to be pretty high. If this is just about Kosovo...

Also, I'm planning to introduce a country-year version which should take care of most overlaps, where multiple CoW codes (for example) map onto an ISO code over different years.

**Comment by LaDilettante** [comment]
Yep, the country year version would seem a lot more useful.


Anh Le

PhD candidate in Political Science, Duke University

________________________________
